walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,147 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Iterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        ArrayList<TimeRange> timesWithOptionals = new ArrayList<>();
+        ArrayList<TimeRange> timesNoOptionals = new ArrayList<>();
+        timesWithOptionals.add(TimeRange.WHOLE_DAY);
+        timesNoOptionals.add(TimeRange.WHOLE_DAY);
+
+        if (request.getDuration() > (24 * 60)) {
+            return new ArrayList<>();
+        } else if (events.size() == 0) {
+            return timesWithOptionals;
+        }
+
+        Collection<String> requestersWithOptionals = new ArrayList<>();
+        requestersWithOptionals.addAll(request.getAttendees());
+        requestersWithOptionals.addAll(request.getOptionalAttendees());
+        Collection<String> requestersNoOptionals = new ArrayList<>();
+        requestersNoOptionals.addAll(request.getAttendees());
+
+        ArrayList<Event> eventsWithOptionals = new ArrayList<>(events);
+        ArrayList<Event> eventsNoOptionals = new ArrayList<>(events);
+
+        ArrayList<Event> currEventList = eventsWithOptionals;
+        Collection<String> currRequesters = requestersWithOptionals;
+        int eventsLength = eventsWithOptionals.size();
+        for (int i = 0; i < 2; i++) {
+            if (i == 1) {
+                currEventList = eventsNoOptionals;
+                currRequesters = requestersNoOptionals;
+                eventsLength = eventsNoOptionals.size();
+            }
+            for (int eventsIndex = 0; eventsIndex < eventsLength; eventsIndex++) {
+                Event currEvent = currEventList.get(eventsIndex);
+                Set<String> currEventAttendees = currEvent.getAttendees();
+                
+                boolean relevantEvent = false;
+                for (String person : currEventAttendees) {
+                    if (currRequesters.contains(person)) {
+                        relevantEvent = true;
+                    }
+                }
+                if (!relevantEvent) {
+                    currEventList.remove(currEvent);
+                    eventsIndex--;
+                    eventsLength--;
+                }
+                
+            }
+        }
+
+        ArrayList<Event> eventsList = eventsWithOptionals;
```

Comment:
I think the event parsing portion of this code and the time parsing portion should be in separate helper methods.